### PR TITLE
Move expired token deletion off validation path

### PR DESCRIPTION
**Pull request by portante**
Work pulled from PR #3114, commits:

 * (26) https://github.com/distributed-system-analysis/pbench/commit/12f0adf73328f9f92c18e4403359f213767f67e7 Token without a user is an internal error
 * (22) https://github.com/distributed-system-analysis/pbench/commit/0e2916009504d992e4ff74d825974cd3dab3ecbe Just remove all expired tokens when looking
 * (12) https://github.com/distributed-system-analysis/pbench/commit/0aa6f90cb80ac167a9a01f9759716a78d20e89a5 s:Move expired token deletion off validation path
 * (08) https://github.com/distributed-system-analysis/pbench/commit/a1c1b94268dbde44dad50c8f12b3317ce301c0cb Move expired token deletion off validation path

Suggested commit message to use below.

----

    The process of validating an API call should NOT incur database deletion
    traffic, where expired tokens could cause unwanted load on the database.
    Instead, only during a new login are ALL expired tokens purged from the
    database.

    The `created` column of `ActiveTokens` is replaced by the new
    `expiration` column to make it easy to find all expired tokens for a
    user and remove them.

    Further, we add a SQLAlchemy `relationship()` between the active token
    table and the user table so that it is trivial to fetch the `User` from
    the `ActiveTokens` object.

    The operation of adding a token for a user is moved to its own method,
    `add_token()`.

### CODE REVIEW

File: lib/pbench/server/api/resources/users_api.py
Comment (webbnh) [inline lib/pbench/server/api/resources/users_api.py:294]: I think text needs some work, since the first sentence seems to be in contradiction with the second half of the second sentence.

I think what we want to say is something like
> If the token is invalid, it won't be in the database; valid but expired tokens may have been removed from the database by other requests; otherwise, the token should be in the database.  If the token is valid, found in the database, and associated with a user, then remove it from the database.

File: lib/pbench/server/api/resources/users_api.py
Comment (webbnh) [inline lib/pbench/server/api/resources/users_api.py:309]: The token is (apparently) valid, and we found it in the database; why do we avoid removing it?  (If it had had a `user`, we _would_ have proceeded with the removal....)

File: lib/pbench/server/api/resources/users_api.py
Comment (webbnh) [inline lib/pbench/server/api/resources/users_api.py:325]: Why bother conditionalizing this?  At this point, it is possible that the token is valid, in the database, associated with a user, but expired...what difference does it make if it is un-expired?

We've already filtered out the case of _invalid_ tokens; I would think the rest of them would be handled the same way at this point.

File: lib/pbench/server/api/resources/users_api.py
Comment (webbnh) [inline lib/pbench/server/api/resources/users_api.py:334]: Now that this change is "done", are we sure that we want to log these?  Logging out with an expired token probably isn't weird.  And, logout should be an audit event, and it's a user action, so should it really be logged (above the `debug` level)?

File: lib/pbench/server/auth/auth.py
Comment (webbnh) [inline lib/pbench/server/auth/auth.py:69]: It would be clearer to state explicitly that the function returns a Tuple (and, then we can say what's in the tuple).

File: lib/pbench/server/database/models/users.py
Comment (webbnh) [inline lib/pbench/server/database/models/users.py:117]: The docstring appears to be stale (or there is some subtlety that I'm missing...):  I don't see any code in this function which removes expired tokens.

Btw, what do lines 112 and 113 do, respectively?  (I.e., why do we need them both?  Or, why do we add `token` to the session instead of adding `self`?)

File: lib/pbench/server/database/models/users.py
Comment (webbnh) [inline lib/pbench/server/database/models/users.py:112]: We don't want to keep some code here which ensures that we don't try to "update" the `auth_tokens` thingie?

File: lib/pbench/test/unit/server/test_user_auth.py
Comment (webbnh) [inline lib/pbench/test/unit/server/test_user_auth.py:172]: Why are we providing _both_ a Bearer token _and_ username/password?  Aren't those redundant credentials?  What does this test beyond what we test at line 181?

File: lib/pbench/test/unit/server/test_user_auth.py
Comment (webbnh) [inline lib/pbench/test/unit/server/test_user_auth.py:228]: We should probably do a query before the login which shows that the tokens that we added are still there...otherwise, the check at line 228 doesn't prove much....

**Review by webbnh — APPROVED** [review]
Here are a bunch of nits and small stuff to consider before you merge this.

File: lib/pbench/server/api/resources/users_api.py
Comment (portante) [reply lib/pbench/server/api/resources/users_api.py:239]: Good question.  I added this statement (I think it was added by me) to debug problems in the unit tests where auth tokens were not being created or created incorrectly.

File: lib/pbench/server/api/resources/users_api.py
Comment (portante) [reply lib/pbench/server/api/resources/users_api.py:297]: I agree with the `debug` level.

File: lib/pbench/server/api/resources/users_api.py
Comment (portante) [reply lib/pbench/server/api/resources/users_api.py:294]: Thanks, I'll reword.

File: lib/pbench/server/api/resources/users_api.py
Comment (portante) [reply lib/pbench/server/api/resources/users_api.py:309]: Actually, the database relationships and constraints should ensure that every token has a user.  So if we have a token without a user that would be an internal error related to a database problem of some sort.

I'll change this code to take advantage of that assertion.

File: lib/pbench/test/unit/server/test_user_auth.py
Comment (portante) [reply lib/pbench/test/unit/server/test_user_auth.py:172]: Tracked in issue #3258.

File: lib/pbench/server/database/models/users.py
Comment (portante) [reply lib/pbench/server/database/models/users.py:112]: We'll just get an exception.

File: lib/pbench/server/database/models/users.py
Comment (portante) [reply lib/pbench/server/database/models/users.py:117]: Yes, the comment is stale, I'll correct.

I'll have to noodle about line 112, I think it is incorrect.

File: lib/pbench/server/api/resources/users_api.py
Comment (portante) [reply lib/pbench/server/api/resources/users_api.py:325]: I'll have to review this more, I am finding it difficult to review the code in the way GitHub is presenting it to me at the moment.

File: lib/pbench/server/database/models/users.py
Comment (portante) [reply lib/pbench/server/database/models/users.py:117]: No this is correct.  Line 112 updates the auth_tokens "column" of the "row" in the users table while the `.add(token)` at line 113 creates the "row" in the active_tokens table for the given token.

File: lib/pbench/test/unit/server/test_user_auth.py
Comment (portante) [reply lib/pbench/test/unit/server/test_user_auth.py:228]: But we just added them at line 218.

File: lib/pbench/server/database/models/active_tokens.py
Comment (dbutenhof) [inline lib/pbench/server/database/models/active_tokens.py:31]: Huh; I hadn't noticed this before. Normally `declarative_base` classes shouldn't have `__init__` methods; they risk inconsistency, because while they *will* be called on an explicit `ActiveTokens(...)` call, they're not *necessary* (the default is a `kwargs` with keywords matching the column names ... and here the mismatch between `auth_token` and `self.token` is a bit annoying), and they're *not* called when SQLAlchemy creates proxies from the DB... which means the temptation to put extra logic into a constructor is potentially confusing. (You can use validator hooks for that, which will be called.)

That is, the only functional difference if this were removed is that you'd need to be careful to `ActiveTokens(token=value, expiration=value)` instead of `ActiveTokens(auth_token=value, expiration=value)`, which wouldn't work because `auth_token` isn't a column. But I don't like the name mismatch anyway. (And, yeah, it's not new with this PR, but I doubt there are that many constructor calls in the code so cleaning it up probably wouldn't be hard...)

File: lib/pbench/server/database/models/users.py
Comment (dbutenhof) [reply lib/pbench/server/database/models/users.py:117]: Note that SQLAlchemy keeps the `relationship` collection and the foreign key in sync automatically. If we made sure the foreign key was assigned, we wouldn't need to manually append to the relationship. (I don't know if it'll complain or overwrite if you append a child to the parent relationship which already has a different foreign key assigned, but of course it's always best not to do both anyway.)

**Review by dbutenhof — DISMISSED** [review]
Some cleanup would be nice, but it doesn't look "evil" 👀

File: lib/pbench/test/unit/server/test_user_auth.py
Comment (webbnh) [reply lib/pbench/test/unit/server/test_user_auth.py:228]: We _requested_ that they be added at line 218, but we don't have any proof that the request had any effect...and then we believe that getting an empty result at line 227 means that the tokens that we "added" were actually removed.

As the code stands, if the assertion at line 228 _doesn't_ fail, there's an alternative explanation:  if could be that the tokens were simply never added at line 218....

File: lib/pbench/server/database/alembic/versions/9df060db17de_tokens_store_expiration.py
Comment (webbnh) [reply lib/pbench/server/database/alembic/versions/9df060db17de_tokens_store_expiration.py:32]: Wasn't this script automatically generated?  @dbutenhof, are you suggesting that we should edit these scripts after generating them?

Comment (dbutenhof) [reply lib/pbench/server/database/alembic/versions/9df060db17de_tokens_store_expiration.py:32]: The script is generated, but usually edited. And I'd be surprised (though it wouldn't be the first time) if Alembic took it upon itself to decide that *renaming* a column was an appropriate upgrade when one column name disappears and another appears! In this case that works with existing data (simply pre-expiring all tokens), but in general it could be a really bad idea...

Comment (webbnh) [reply lib/pbench/server/database/alembic/versions/9df060db17de_tokens_store_expiration.py:32]: I agree:  if we're entitled to edit this script, then we should be removing and creating (i.e, wiping) these columns instead of renaming them.

File: lib/pbench/server/api/resources/users_api.py
Comment (npalaska) [inline lib/pbench/server/api/resources/users_api.py:241]: Why do it during the login? I don't know the impact of it during the login process but I think the better place to remove all the expired tokens would be in the logout. 

I might also suggest if we are removing the expired tokens during the login we might do it only for the user who is trying to log in instead of the _Any_ user.

**Review by npalaska — DISMISSED** [review]
I think this generally looks good, I agree with some of the comments already made here and I also have a comment.

File: lib/pbench/server/api/resources/users_api.py
Comment (webbnh) [reply lib/pbench/server/api/resources/users_api.py:241]: We want this operation on a common but low frequency path.  Login is a nice reliable spot.  I don't see how logout would be better, and, for many users, they will probably just close the tab and never log out, so that wouldn't be a particularly reliable point in the flow.

As for removing expired tokens, an expired token is invalid no matter who it belongs to, and we can get them all with a single database operation, so why filter it to only the current user?

Comment (dbutenhof) [reply lib/pbench/server/api/resources/users_api.py:241]: The discussion is nearly moot as we expect our API keys to be permanent (at least until manually revoked through some API) and we don't hold or control Auth tokens from Keycloak. So this code has a short lifetime but might be valuable for our initial March deployment.

File: lib/pbench/server/api/resources/users_api.py
Comment (portante) [reply lib/pbench/server/api/resources/users_api.py:325]: Given https://github.com/distributed-system-analysis/pbench/pull/3231#discussion_r1107091136, I think this can be removed.

File: lib/pbench/test/unit/server/test_user_auth.py
Comment (portante) [reply lib/pbench/test/unit/server/test_user_auth.py:228]: Actually, we need to commit that transaction.

File: lib/pbench/server/database/alembic/versions/9df060db17de_tokens_store_expiration.py
Comment (portante) [reply lib/pbench/server/database/alembic/versions/9df060db17de_tokens_store_expiration.py:32]: I chose to use alter column instead of drop and create.  It seems silly to do anything about the table data since no token will be "created" in the future, and the first login will take care of the expirations.

File: lib/pbench/server/api/resources/users_api.py
Comment (portante) [reply lib/pbench/server/api/resources/users_api.py:251]: @dbutenhof we use this same sequence in `datasets.py`.  Can you fix it there and I'll fix this one?

File: lib/pbench/server/api/resources/users_api.py
Comment (portante) [reply lib/pbench/server/api/resources/users_api.py:241]: People must login to do something, but they more often than not fail to logout.

File: lib/pbench/server/database/models/active_tokens.py
Comment (portante) [reply lib/pbench/server/database/models/active_tokens.py:31]: I'll fix this.

File: lib/pbench/test/unit/server/test_user_auth.py
Comment (portante) [reply lib/pbench/test/unit/server/test_user_auth.py:228]: How could they "simply never be added at line 218"?  That is what `user.add_token()` method is supposed to do.  Why would this code, as opposed to any other code that calls the same method, need to verify they were added?

This is testing user auth code flow, not the database.  If the SQLAlchemy ORM fails to commit those rows to the database, we have a much larger problem.

File: lib/pbench/server/api/resources/users_api.py
Comment (dbutenhof) [reply lib/pbench/server/api/resources/users_api.py]: Ha; maybe elsewhere, too, then. Ah, well.

Comment (portante) [reply lib/pbench/server/api/resources/users_api.py]: Those were the only two I could find.

**Comment by portante** [comment]
Bugger all, CI jobs won't execute if they have conflicts with the target branch.

File: lib/pbench/test/unit/server/test_user_auth.py
Comment (webbnh) [reply lib/pbench/test/unit/server/test_user_auth.py]: This is unit test code...it runs in a weird environment in which reality is regularly warped.

The assertion at line 228 is supposed to detect that the items have been removed from the database...why not perform the inverse check before invoking the CUT to ensure that the expected initial state is in place?  This ensures that the transition that we're trying to observe actually takes place.  ("Now we see them, now we don't -- the CUT is working.")

I don't think that it's a big deal to add the assertion (or that it will be costly to execute).  You evidently do, so I'll defer to you and close this conversation.

**Review by webbnh — APPROVED** [review]
GTG

**Comment by portante** [comment]
Bugger all, I messed up the alembic migration in the conflict merge.

**Review by webbnh — APPROVED** [review]
Looks fine AFAICT.

**Pull request merged**
